src/document.rs

```diff
@@ -129,172 +172,447 @@ pub struct Prescription<'a> {
-/// metadata linked to the related node.
-fn subsume_comments(document: Document) -> HowserResult<Document> {
-    {
-        let traverser = document
-            .root
+        if let Some(l2_node) = l1_node
             .capabilities
             .traverse
             .as_ref()
-            .ok_or(HowserError::CapabilityError)?;
-
-        for (node, event) in traverser.iter() {
-            let getter = node.capabilities
-                .get
-                .as_ref()
-                .ok_or(HowserError::CapabilityError)?;
-
-            match (event, getter.get_type()?) {
-                (IterEventType::Exit, NodeType::CMarkNodeHtmlInline)
-                | (IterEventType::Enter, NodeType::CMarkNodeHtmlInline) => {
-                    subsume_comment(&node, &document)?;
+            .ok_or(HowserError::CapabilityError)?
+            .next_sibling()?
+        {
+            if types_match(&l1_node, &l2_node)? {
+                match l1_node
+                    .capabilities
+                    .get
+                    .as_ref()
+                    .ok_or(HowserError::CapabilityError)?
+                    .get_type()?
+                {
+                    NodeType::CMarkNodeHeading | NodeType::CMarkNodeItem => {
+                        if let Some(l3_node) = l2_node
+                            .capabilities
+                            .traverse
+                            .as_ref()
+                            .ok_or(HowserError::CapabilityError)?
+                            .next_sibling()?
+                        {
+                            if types_match(&l1_node, &l3_node)? {
+                                l3 = LookaheadType::new(Some(l3_node))?;
+                            } else {
+                                l3 = LookaheadType::Other(Some(l3_node));
+                            }
+                        }
+                    }
+                    _ => {
+                        l3 = LookaheadType::Other(l2_node
+                            .capabilities
+                            .traverse
+                            .as_ref()
+                            .ok_or(HowserError::CapabilityError)?
+                            .next_sibling()?);
+                    }
                 }
-                _ => continue,
+                l2 = LookaheadType::new(Some(l2_node))?;
+            } else {
+                l2 = LookaheadType::Other(Some(l2_node));
             }
         }
+        let l1 = LookaheadType::new(Some(l1_node))?;
+
+        debug!("Lookahead Sequence: {:?}, {:?}, {:?}", l1, l2, l3);
+
+        match (l1, l2, l3) {
+            (
+                LookaheadType::DiscreteAnnotated((annotation, match_type)),
+                LookaheadType::DiscreteLiteral(target),
+                LookaheadType::Ditto(ditto)
+            ) => {
+                process_child_block_elements(&target, document)?;
+                document.set_match_type(&target, match_type)?;
```

Comment:
If this is doing what it appears to be doing, then `document` should be a `&mut` here, as should the `self` param to set_match_type.

Reply:
The reason `document` is not mut is that the mutations are happening across the FFI. There will be compiler warnings if I mark it mutable. As for `set_match_type`, I think I had a reason for using the interior mutability pattern there, but I can't remember right now. I'll have to think about it.